dds/DCPS/DataReaderImpl_T.h

```diff
@@ -738,23 +685,21 @@ namespace OpenDDS {
     const bool filter_has_non_key_fields =
       evaluator.has_non_key_fields(getMetaStruct<MessageType>());
 
-    for (SubscriptionInstanceMapType::iterator iter = instances_.begin(), end = instances_.end(); iter != end; ++iter) {
-      SubscriptionInstance& inst = *iter->second;
+    const HandleSet& matches = lookup_matching_instances(sample_states, view_states, instance_states);
+    for (HandleSet::const_iterator it = matches.begin(); it != matches.end(); ++it) {
```

Comment:
Can the state of the instance change in this loop?  If so, it may invalidate the iterators.